### ISSUE TITLE
Error is thrown when setting fillcharsAdd a short description of the bug here

### ISSUE THREAD

**Issue by abhiyandhakal** [LunarWatcher/auto-pairs#78]
**OS** (name + version; i.e. Linux Mint 20, Windows 10 2004, Macos Big Sur, etc.):
Arch Linux

**What vim? (+ version)** (i.e. vim, gvim, nvim, nvim-qt, mvim, ...):
NVIM 0.9.0

**Reproduced in other variants of Vim?** (Optional):

**Autopairs config** (if applicable):
I have just installed autopairs using packer, no other config

**Describe the bug**

I have set fillchars in my neovim config as follows:
![Screenshot_20230515_072453](https://github.com/LunarWatcher/auto-pairs/assets/96732471/d4038eb9-2f5c-4f27-97db-bf0b8ae0e690)

Following error is thrown when I enter the command `:vnew`:
![Screenshot_20230515_072626](https://github.com/LunarWatcher/auto-pairs/assets/96732471/7fd830ab-7872-4222-b2cd-eafd08a8c864)

After this error comes along, telescope find files also stops working.

**Steps to reproduce**

Install autopairs and then set fillchars.

**Comment by abhiyandhakal** [comment]
The bug still remains once I remove the setting of fillchars

**Comment by LunarWatcher** [comment]
What's the output of `:verbose imap <cr>`?

**Comment by LunarWatcher** [comment]
Downloaded and used your dotfiles, can't reproduce it there. <CR> maps to nvim-cmp, but it appears to map compatibly. 

In either case, fillchars is irrelevant. There's either another plugin interfering, or my best guess seeing as reports of this bug so far have only come from nvim, nvim might've broken maparg. According to `:h maparg`:
```
		When {dict} is there and it is |TRUE| return a dictionary
		containing all the information of the mapping with the
		following items:			*mapping-dict*
		  "lhs"	     The {lhs} of the mapping as it would be typed
		  "lhsraw"   The {lhs} of the mapping as raw bytes
		  "lhsrawalt" The {lhs} of the mapping as raw bytes, alternate
			      form, only present when it differs from "lhsraw"
		  "rhs"	     The {rhs} of the mapping as typed.
```

Since dict is set to true, rhs is defined. This is also backed up further by the docs' statement for what's returned when the mapping doesn't exist:
```
		When there is no mapping for {name}, an empty String is
		returned if {dict} is FALSE, otherwise returns an empty Dict.
		When the mapping for {name} is empty, then "<Nop>" is
		returned.
```

Which is already covered with an `if empty(info)` block in the code (Keybinds.vim#215 at the current commit, arsed linking to it). However, since I can't reproduce and consequently can't debug, I need more information (and preferably a minimal config file that's able to reproduce the problem).

**Comment by abhiyandhakal** [comment]
You are right, it might not be the issue of fillchars.

I removed fillchars, the config, and all the cache including the installed plugins in `~/.local/share/nvim` and then downloaded and set up my config again. The error still persists. But when I removed the package `LunarWatcher/auto-pairs`,  the error is no longer thrown.

I am new to neovim. So I don't understand the errors thrown, so I have only referred to those as error.

<hr>

This is the output of `:verbose imap <cr>`:
```
i  <CR>        * <Lua 184: ~/.local/share/nvim/site/pack/packer/start/nvim-cmp/lua/cmp/utils/keymap.lua:127>                                                                                                                               
                 cmp.utils.keymap.set_map                                                                                                                                                                                                  
        Last set from Lua   
```

**Comment by abhiyandhakal** [comment]
Here is my [neovim config](https://github.com/abhiyandhakal/nvim-config).

**Comment by LunarWatcher** [comment]
I tried with your dotfiles yesterday, and still couldn't reproduce. Tried again now specifically with vsplit, and still can't reproduce. `:verbose imap <cr>` is as expected. Tried on both 0.9.0 and 0.10.0-dev

I'd recommend disabling other plugins to see if disabling another plugin also fixes it. That would check if there's an incompatibility, even if it's an obscure one. If there isn't, I'll personally be blaming nvim (the function call is made and handled according to spec, and if there isn't another plugin interfering, that only leaves one common factor between this and the other case; nvim itself), and recommend you open an issue there instead

**Comment by abhiyandhakal** [comment]
Thank you for the suggestion. I checked out with all the plugins. But the problem was with the configuration provided in [lsp-zero](https://github.com/VonHeikemen/lsp-zero.nvim). I didn't know that the docs was updated. Turns out this was the problem:
```
local lsp = require('lsp-zero').preset({
	name = 'minimal',
	set_lsp_keymaps = true,
	manage_nvim_cmp = true,
	suggest_lsp_servers = false,
})
```
The error is no longer thrown if it is replaced by:
```
local lsp = require('lsp-zero').preset({})
```
I still don't know what caused the issue... will look into it more...Thank you for your time :)

**Comment by LunarWatcher** [comment]
```
	manage_nvim_cmp = true,
```
is presumably why nvim-cmp maps `<cr>`, but that still doesn't make sense. Why would nvim-cmp mapping `cr` break `maparg`? The other weird part is that I can't trigger the bug manually even with cr mapped, which doesn't make sense

**Comment by abhiyandhakal** [comment]
One of my friends is using my config, and he gets the error, but sometimes only. It sometimes throws no error but throws the same error other times in his laptop

**Comment by abhiyandhakal** [comment]
So is it the problem of nvim_cmp or...?

**Comment by LunarWatcher** [comment]
Can you switch to the 78-debug branch and try triggering the error? It consists of a single commit (well, two, the second one just adds a string in front of the dict) echoing the contents of the `info` dict. It's not going to fix anything, but it'll show what `maparg` actually returns. That's theoretically the last missing piece before blame can be assigned. Should appear above the error messages. If it doesn't or you don't see it, clear the errors and check `:messages`

(You can (and should) switch back to the master branch afterwards. It's a temporary branch purely for debug)

**Comment by abhiyandhakal** [comment]
```
maparg returned {'lnum': 0, 'script': 0, 'mode': 'i', 'desc': 'cmp.utils.keymap.set_map', 'callback': function('<lambda>4'), 'noremap': 1, 'lhs': '<CR>', 'lhsraw': '^M', 'nowait': 0, 'expr': 0, 'sid': -8, 'buffer': 0, 'silent': 1}
```
This is what returned when triggering the error.

**Comment by LunarWatcher** [comment]
Sigh... That confirms it's nvim's fault. I'll open an issue in their repo, though I doubt there's a fix for it. Doc update is probably the best-case here. I'll implement a hack, but the short solution appears to be disregarding Lua-based maps when mapping compatibly. That's annoying

**Comment by LunarWatcher** [comment]
As for why it isn't consistently reproducible everywhere, my best guess is the plugin init order, though that still doesn't a ccount for everything. 

It's perfectly possible that what I'm seeing is a result of auto-pairs loading first, and nvim-cmp loading second, while what you're seeing is nvim-cmp loading first, and auto-pairs choking when mapping compatibly due to nvim apparently not defining rhs for lua-based mappings. I have no clue why this would happen though, but I'm also not deep enough in (n)vim code to tell how plugin load order works.

**Comment by LunarWatcher** [comment]
No, auto-pairs isn't Lua-based. Even then, auto-pairs mostly operates on a buffer level. The order difference is realistically going to be in autocmd call order, and I have no clue how you affect it. Some old comments by jiangmiao led me to believe plugins were loaded alphabetically, but that would mean **a**uto-pairs is always loaded before **n*vim-cmp. I'm not sure what exactly the loading order is though. It could be something else, or even platform-dependent. 

The only good workaround here is defining your own map, manually combining auto-pairs with whatever else you have. I'm pretty sure I wrote docs on how to do that, though that's Vimscript-based, and assumes you're able to independently call all the things you want to call. Based on what I saw of nvim-cmp's code, I doubt you'd be able to map nvim-cmp to a separately invocable function. I'll have to look at auto-pairs' options for dealing with it later  though. The solutions also depend on what direction neovim goes in, which first requires me to finish the bug report

**Comment by abhiyandhakal** [comment]
I think disabling `AutoPairsMapCR` will do for me. I really cannot work with `<C-y>`. I can press enter twice. Even if I don't, lsp's formatting does it. So I think I will go with option 1

**Comment by abhiyandhakal** [comment]
I will also try to combine the functions... If I am able to do so, then I will post it here.

**Comment by thesergiomiguel** [comment]
Disabling the global `MapCR` and enabling the the buffer local one with `BufEnter` works for me and seems to be the path with the least side effects.

**Comment by LunarWatcher** [comment]
Due to [AI slop in vim](https://hachyderm.io/@AndrewRadev/116175986749599825) (and in neovim both by extension and independently), I have moved to emacs with evil mode. This plugin is no longer maintained, so this will no longer be fixed (though this particular bug still hinged on an upstream bug in neovim that was reported nearly 3 years ago, so pretty sure reliable neovim support always was a lost cause).
